File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-7.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        if (s.length() == 1) {
            return 1;
        }

        int maxLength = 0;
        int l = 0;
        Map<Character, Integer> map = new HashMap<>();
        for (int r = 0; r < s.length(); r++) {
            char c = s.charAt(r);
            if (map.containsKey(c)) {
                int oldIndex = map.get(c);
                l = Math.max(oldIndex + 1, l);
            }

            map.put(c, r);
            maxLength = Math.max(maxLength, r - l + 1);
        }

        return maxLength;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-7.js
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s, k) {
        const map = {};

        let left = 0;
        let right = 0;
        let maxLength = 0;

        while (right < s.length) {
            const c = s.charAt(right);

            if (c in map) {
                map[c] = 1 + map[c];
            } else {
                map[c] = 1;
            }

            while ((right - left + 1) - Math.max(...Object.values(map)) > k) {
                const leftC = s.charAt(left);
                map[leftC] = map[leftC] - 1;
                left++;
            }

            maxLength = Math.max(maxLength, right - left + 1);
            right++;
        }

        return maxLength;
    }
}
